sklearn/ensemble/stacking.py

```diff
@@ -0,0 +1,186 @@
+"""Stacking API"""
+
+# Author: Caio Oliveira <caioaao@gmail.com>
+# License: BSD 3 clause
+
+from ..base import (BaseEstimator, TransformerMixin, MetaEstimatorMixin)
+from ..model_selection import cross_val_predict
+from ..pipeline import (_name_estimators, FeatureUnion, Pipeline)
+from ..preprocessing import FunctionTransformer
+
+
+class StackMetaEstimator(BaseEstimator, MetaEstimatorMixin, TransformerMixin):
+    """Transformer to turn estimators into meta-estimators for model stacking
+
+    In stacked generalization, meta estimators are combined in layers to
+    improve the final result. To prevent data leaks between layers, a procedure
+    similar to cross validation is adopted, where the model is trained in one
+    part of the set and predicts the other part. In `StackMetaEstimator`, it
+    happens during `fit_transform`, as the result of this procedure is what
+    should be used by the next layers.
+
+    Parameters
+    ----------
+    base_estimator : the estimator to be blended.
+
+    cv : cv to be used, optional (default=3)
+        Will be passed to `cross_val_predict` during `fit_transform`.
+
+    method : string, optional (default='auto')
+        Invokes the passed method name of the passed estimator. If the method
+        is `auto`, will try to invoke `predict_proba` or `predict` in that
+        order.
+
+    n_jobs : int, optional (default=1)
+        Number of jobs to be passed to `cross_val_predict` during
+        `fit_transform`.
+
+    """
+    def __init__(self, base_estimator, cv=3, method='auto', n_jobs=1):
+        self.base_estimator = base_estimator
+        self.cv = cv
+        self.method = method
+        self.n_jobs = n_jobs
+
+    def fit(self, *args, **kwargs):
+        raise NotImplementedError("Use `fit_transform` instead")
+
+    def _method_name(self):
+        if self.method == 'auto':
+            if getattr(self.base_estimator, 'predict_proba', None):
+                method = 'predict_proba'
+            else:
+                method = 'predict'
+        else:
+            method = self.method
+
+        return method
+
+    def transform(self, *args, **kwargs):
+        t = getattr(self.base_estimator, self._method_name())
+        preds = t(*args, **kwargs)
+
+        if preds.ndim == 1:
+            preds = preds.reshape(-1, 1)
+
+        return preds
+
+    def fit_transform(self, X, y, **fit_params):
+        preds = cross_val_predict(self.base_estimator, X, y, cv=self.cv,
+                                  method=self._method_name(),
+                                  n_jobs=self.n_jobs, fit_params=fit_params)
+
+        self.base_estimator.fit(X, y, **fit_params)
+
+        if preds.ndim == 1:
```

Comment:
it doesn't flatten multi-output: it's just transforming 1d arrays into a column vector so we can hstack them together